src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp

```diff
@@ -548,20 +565,39 @@ Pipe ReadFromSystemNumbersStep::makePipe()
         return pipe;
     }
 
+    const auto end = std::invoke(
+        [&]() -> std::optional<UInt64>
+        {
+            if (numbers_storage.limit.has_value())
+                return *(numbers_storage.limit) + numbers_storage.offset;
+            return {};
+        });
+
     /// Fall back to NumbersSource
+    /// Range in a single block
+    const auto block_range = max_block_size * numbers_storage.step;
```

Comment:
Let's add explicit type for all variables here to assure that we don't have overflow (and write comments, please, why overflow is not possible here)

Reply:
I dont see how explicit types would help with the overflow. E.g. `const UInt64 block_range = ...` won't help in case the result of the multiplication is bigger than `2^64`. I think we either have to use `mulOverflow` or cast the number to `UInt128` or even `UInt256`.

Do I miss something? If not, then I think handling the overflow is out of the scope of this PR, because it could happen even before. I am happy to take a look at it, but wouldn't like to mix an improvement with a bugfix if not necessary.